blog/_posts/2017-12-05-ml&pl.md

```diff
@@ -0,0 +1,88 @@
+
+In fact, you can program in "lazy" TensorFlow style in any language. Consider the following JavaScript code, which implements a trivial function (`add`) in this style:
+
+{% highlight julia %}
+function add(a,b) {
+  return `${a}+${b}`;
+}
+x = 1; y = 2
+z = add(‘x’, ‘y’) // ‘x+y’
+eval(z) // 3
+
+x = 4
+eval(z) // 6
+{% endhighlight %}
+
+Here we are *metaprogramming* – writing code that writes code. In this case both the meta-language and the target language are the same (JavaScript) but they could just as well be different languages (as in the C preprocessor for the C language), or we can use a data structure (an [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree)) instead of a string – the principle is the same. In TensorFlow, Python serves as a meta-language for writing programs in TF’s graph-based language[^ast]. If you’re not convinced, consider that TensorFlow’s graph even supports constructs like variable scoping and control flow – but rather than using Python syntax, you manipulate these constructs through an API.
+
+[^ast]: TensorFlow’s graph is effectively a dataflow-based AST (Abstract Syntax Tree).
+
+TensorFlow and similar tools present themselves as "just libraries", but they are extremely unusual ones. Most libraries provide a simple set of functions and data structures, not an entirely new programming system and runtime. Why is such a complex approach necessary?
+
+## Why create a new language?
+
+The core reason for building new languages is simple: ML research has extremely high computational demands, and simplifying the modelling language makes it easier to add domain-specific optimisations and features. Training models requires excellent hardware support, good numerics, low interpreter overhead and multiple kinds of parallelism, all of which general-purpose languages like Python struggle to provide.
+
+There’s a snag, though. This approach relies on simplifying assumptions (ML models won’t be recursive, or need custom gradients, right?), which make it easier to apply optimisations or deploy to small devices. Unfortunately for engineers, model complexity has increased and researchers thoroughly enjoy violating these assumptions. Models now demand conditional branching (ok, easy enough to hack in), loops for recurrence (less easy but possible), and even [recursion over trees](https://arxiv.org/pdf/1503.00075.pdf) (virtually impossible to deal with). In many areas of ML, including [neural networks](https://blog.keras.io/the-future-of-deep-learning.html) and [probabilistic programming](https://eng.uber.com/pyro/), models are becoming increasingly like programs, including ones that reason about *other* programs (e.g. [program generators](https://arxiv.org/pdf/1705.03633.pdf) and [interpreters](https://arxiv.org/abs/1605.06640)). They often include non-differentiable components like Monte Carlo Tree Search, making it ever more challenging to build runtimes that are general enough while achieving top performance.
```

Comment:
This can be less declarative. The "what" is there, but not the "why". Why is recursion hard? Why is tree recursion then almost impossible? Maybe a picture showing how the computational graphs grow exponentially?

Reply:
Well the why is simple – TF doesn't support recursion, because it made a simplifying assumption that it wouldn't be needed. Isn't that stated above?

Reply:
Is it? I don't remember it explicitly saying that TF cannot do it, and I didn't know it couldn't. I assumed it could, but it would be slow to generate the graph. I just did a Ctrl+f and didn't find a mention of it.

Reply:
> This approach relies on simplifying assumptions (ML models won’t be recursive, or need custom gradients, right?)

Reply:
I didn't realize that statement meant TensorFlow cannot do it. I thought it meant that TF cannot optimize it well.

Reply:
Would replacing "simplifying the modelling language" with something like "restricting what the language can express" above make this limitation clearer?